### PR TITLE
Improvements, comments and harmonization of the code

### PR DESCRIPTION
**Pull request by ClaudeCoulombe**
Greetings Mr Liashchynskyi,

I know, that's a pretty old piece of code but I've found a small issue. In fact there are some incompatibilities between backprop code and the activation functions code. I've added some comments to explain and add more appropriate credits. The code as formulated before will never converge to a solution for XOR since in some cases the activation is applied two times. 

You've just to check it and merge. 
It's up to you.

Thanks again for sharing you code on GitHub.

Claude COULOMBE
Montréal

### CODE REVIEW

File: nn.js [hunk 1]
```diff
@@ -45,27 +49,33 @@ export class NeuralNetwork {
...
-
+            let hidden_layer_linear = multiply(input_layer, this.synapse0)
+            let hidden_layer_activated = hidden_layer_linear.map(v => this.activation(v, false));
+            let output_layer_linear = multiply(hidden_layer_activated, this.synapse1)
+            let output_layer_activated = output_layer_linear.map(v => this.activation(v, false));
+            // Compute output error
             let output_error = subtract(target, output_layer);
-            let output_delta = dotMultiply(output_error, output_layer.map(v => this.activation(v, true)));
+            // Backpropagation of the output error to the output layer 
+            let output_delta = dotMultiply(output_error, multiply(hidden_layer_activated, this.weight1).map(v => this.activation(v, true)));
```
Comment (liashchynskyi) [inline nn.js]: What is `this.weight1`?

File: nn.js [hunk 2]
```diff
@@ -1,6 +1,7 @@
-import {random, multiply, dotMultiply, mean, abs, subtract, transpose, add} from 'mathjs'
+import {random, multiply, dotMultiply, mean, abs, subtract, transpose, add} from './mathjs'
```
Comment (liashchynskyi) [inline nn.js]: It's a package. There's is no file called `mathjs`

**Review by liashchynskyi — CHANGES_REQUESTED** [review]

File: nn.js [hunk 3]
```diff
@@ -45,27 +49,33 @@ export class NeuralNetwork {
...
             let input_layer = input;
-            let hidden_layer = multiply(input_layer, this.synapse0).map(v => this.activation(v, false));
-            let output_layer = multiply(hidden_layer, this.synapse1).map(v => this.activation(v, false));
-
+            let hidden_layer_linear = multiply(input_layer, this.synapse0)
+            let hidden_layer_activated = hidden_layer_linear.map(v => this.activation(v, false));
+            let output_layer_linear = multiply(hidden_layer_activated, this.synapse1)
+            let output_layer_activated = output_layer_linear.map(v => this.activation(v, false));
+            // Compute output error
             let output_error = subtract(target, output_layer);
```
Comment (liashchynskyi) [inline nn.js]: `output_layer` does not exist anymore. Did you even run this code?

File: nn.js (hunk 1 above)
Comment (liashchynskyi) [inline nn.js]: Why do you multiple `hidden_layer_activated` two times? This is already activated, just get derivative from it

**Comment by liashchynskyi** [comment]
@ClaudeCoulombe I read this article you add in comments. Seems that derivative is wrong. Sigmoid derivative should be `sigmoid(x) * (1 - sigmoid(x))` NOT just `x*(1-x)`. Why everyone writes it down like this (I mean the second way)...

File: nn.js (hunk 1 above)
Comment (ClaudeCoulombe) [reply nn.js]: Oups! 
Sorry, that were not the good files...

**Comment by ClaudeCoulombe** [comment]
Sorr Mr Liashchynskyi ,

I've just tested it with node.js

> node index.js       
Error: 0.5000137685734031
Error: 0.016143712642404932
Error: 0.01107887514649991
Error: 0.008944596826788258
Error: 0.007698572294691791
Matrix {
  _data: [
    [ 0.00790482578515504 ],
    [ 0.9942857212818826 ],
    [ 0.9927391774120087 ],
    [ 0.006553553104046777 ]
  ],
  _size: [ 4, 1 ],
  _datatype: undefined
}

**Comment by liashchynskyi** [comment]
@ClaudeCoulombe Ok, thanks for your contribution! I'll review and merge this later.

**Pull request closed without merging**